### PR TITLE
add tmp_upath_factory fixture

### PR DESCRIPTION
**Pull request by dtrifiro**

### CODE REVIEW

File: src/pytest_servers/fixtures.py [hunk 1]
```diff
@@ -59,3 +63,18 @@ def tmp_upath(request):
     elif param == "s3":
         return request.getfixturevalue("s3path")
     raise ValueError(f"unknown {param=}")
+
+
+@pytest.fixture
+def make_tmp_upath(request: "FixtureRequest"):
```
Comment (efiop) [inline src/pytest_servers/fixtures.py:69]: How about going with the factory to follow `tmp_path_factory` and `tmp_path`?

File: src/pytest_servers/fixtures.py [hunk 2]
```diff
@@ -14,16 +18,16 @@ def aws_region_name():
 
 @pytest.fixture
 def s3_server(monkeypatch):
-    """Spins up a moto s3 server on 127.0.0.1:5555."""
+    """Spins up a moto s3 server on a random port."""
     monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "foo")
     monkeypatch.setenv("AWS_ACCESS_KEY_ID", "foo")
-    with MockedS3Server() as server:
+    with MockedS3Server(random.randint(60000, 65535)) as server:
         yield server
```
Comment (efiop) [inline src/pytest_servers/fixtures.py:25]: In dvc we use one single moto server (to reflect single real s3 server) and use tmp paths within it. Spamming ports with a fixture of such a low scope might become problematic with a growing number of tests.

File: src/pytest_servers/fixtures.py [hunk 3]
```diff
@@ -59,3 +63,18 @@ def tmp_upath(request):
...
+
+@pytest.fixture
+def make_tmp_upath(request: "FixtureRequest"):
+    """Factory function for temporary directories on different filesystems."""
+
+    def _make_tmp(fs: str = "local"):
+        if fs == "local":
+            return request.getfixturevalue("local_path")
+        elif fs == "s3":
+            return request.getfixturevalue("s3path")
```
Comment (efiop) [inline src/pytest_servers/fixtures.py:76]: If I understand correctly, with a default scope of those fixtures, calling `make_tmp_upath` twice will give you the same result, which defeats the purpose of the factory.

File: src/pytest_servers/s3.py [hunk 4]
```diff
@@ -23,7 +23,11 @@ def __enter__(self):
             if r.ok:
                 raise RuntimeError("moto server already up")
         self.proc = subprocess.Popen(
-            shlex.split("moto_server s3 -p %s" % self.port)
+            shlex.split(
+                f"moto_server s3 -p {self.port}",
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
```
Comment (efiop) [inline src/pytest_servers/s3.py:30]: Looks like we are seriously polluting your machine ports, which will turn ugly. Take a look at how we do it in dvc - we spawn one container with moto (and azurite in the future, for example) to simulate real s3 and we use a random single port without polluting.

Comment (efiop) [reply src/pytest_servers/s3.py:30]: https://github.com/iterative/dvc-s3/tree/main/dvc_s3/tests

Comment (efiop) [reply src/pytest_servers/s3.py:30]: If we at least don't spam lots of servers - this will become minor 🙂

File: src/pytest_servers/s3.py [hunk 5]
```diff
@@ -36,3 +44,44 @@ def close(self):
...
+        with monkeypatch_session.context() as m:
+            try:
+                m.delenv("AWS_PROFILE")
+            except KeyError:
+                pass
+            m.setenv("AWS_ACCESS_KEY_ID", "pytest-servers")
+            m.setenv("AWS_SECRET_ACCESS_KEY", "pytest-servers")
+            m.setenv("AWS_SECURITY_TOKEN", "pytest-servers")
+            m.setenv("AWS_SESSION_TOKEN", "pytest-servers")
+            m.setenv("AWS_DEFAULT_REGION", "eu-south-1")
```
Comment (efiop) [inline src/pytest_servers/s3.py:73]: Seems different from `aws_region` fixture

Comment (dtrifiro) [reply src/pytest_servers/s3.py:73]: Got rid of this, we'll add it back it it creates any trouble

File: src/pytest_servers/s3.py (hunk 4 above)
Comment (dtrifiro) [reply src/pytest_servers/s3.py:34]: Added an optional flag (`--moto-server`) to set a custom port. In #13 I introduce a docker fixture which can be used to spawn docker containers. I will introduce some proper port management after that

**Review by efiop — APPROVED** [review]

**Pull request merged**